pkg/asset/machines/master.go

```diff
@@ -114,9 +138,110 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 	default:
 		return fmt.Errorf("invalid Platform")
 	}
+
+	machines := []interface{}{}
+	if len(m.Machines) == 0 {
+		for _, machine := range m.MachinesDeprecated {
+			machines = append(machines, machine)
+		}
+	} else {
+		for _, machine := range m.Machines {
+			machines = append(machines, machine)
+		}
+	}
+
+	for i, machine := range machines {
+		data, err := yaml.Marshal(machine)
+		if err != nil {
+			return errors.Wrapf(err, "marshal master %d", i)
+		}
+
+		m.FileList = append(m.FileList, &asset.File{
+			Filename: filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, strconv.Itoa(i))),
+			Data:     data,
+		})
+	}
+
+	asset.SortFiles(m.FileList)
+
 	return nil
 }
 
+// Files returns the files generated by the asset.
+func (m *Master) Files() []*asset.File {
+	return m.FileList
+}
+
+// Load reads the asset files from disk.
+func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
+	file, err := f.FetchByName(filepath.Join(directory, MasterUserDataFileName))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	m.FileList = []*asset.File{file}
+
+	m.Machines = []machineapi.Machine{}
+	m.MachinesDeprecated = []clusterapi.Machine{}
+	fileList, err := f.FetchByPattern(filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, "*")))
```

Comment:
We should return `false` if there are no master machine files since we require at least one master machine.

Reply:
> We should return `false` if there are no master machine files since we require at least one master machine.

If we find a master user-data file but no master machines, we currently return "succesfully loaded" here (because we did successfully load part of the asset), and then later on error out over [here][1], to avoid `[0]` access panics.  I think we want to keep that guard, so we don't *need* another guard here.

If you're trying to trigger a regeneration instead of error out, I think we'd need to split master user-data out into it's own asset.  I'm fine with that (#556 takes this to its inevitable, asset-for-every-file conclusion), but I'd rather punt on it for this PR ;).

[1]: https://github.com/wking/openshift-installer/blob/d5c01c0ef2d016b30ba750e97bc17db96f30f129/pkg/asset/cluster/tfvars.go#L102-L103

Reply:
My preference is that each asset enforces its out invariants. It is an invariant of the `machines.Master` asset that there is at least one master. That asset should be the one that is complaining when the invariant is not met. I am fine with returning an instead of returning `false`, but I would like this `Load` function to be the one returning the error.

Reply:
Compare it to how loading works for the `InstallConfig` asset. If the user sets the master replicas to 0 in `install-config.yaml`, then loading the `InstallConfig` asset fails. The installer does not wait until some later asset tries to use the replica count to decide that a value of 0 is not valid.

Reply:
> It is an invariant of the `machines.Master` asset that there is at least one master. That asset should be the one that is complaining when the invariant is not met. I am fine with returning an instead of returning false, but I would like this `Load` function to be the one returning the error.

Guards (like "`len(mySlice) > 0`?") that live far from the action they're guarding (like `mySlice[0]`) are only reliable when you stay on top of the call-chain to ensure all access to the guarded code passes through the guard.  I'm fine with belt-and-suspenders guards in both places, but I don't think we should drop the local guards that live right before the guarded action.  I've pushed 955cf95 -> 1e32e76 adding a `Master.Load` guard which will generate error messages like:

```
FATAL failed to fetch Bootstrap Ignition Config: failed to load asset "Master Machines": master machine manifests are required if you also provide openshift/99_openshift-cluster-api_master-user-data-secret.yaml 
```

Reply:
I can live with that.

My opinion on the matter is that the later error is a code error rather than a user error. Whether the installer panics at that point or displays a less jarring error, it is still a failure if the user sees it. I don't see much value in carrying around the code for the extra guard to catch a code error. The guard is not even sufficient to protect against running over the bounds of the slice because there is no guarantee that `masterCount` came from length of the slice being accessed. It could be that `MachinesDeprecated` had a positive length while `Machines` had a zero length, even though the platform is AWS.